### PR TITLE
Feature/initial reference docs

### PR DESCRIPTION
**Pull request by tjcrowder**
Doesn't go into use cases or patterns/anti-patterns yet.

### CODE REVIEW

File: reference.md [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+    * [`FinalizationRegistry.prototype.register`](#finalizationregistryprototyperegister)
+    * [`FinalizationRegistry.prototype.unregister`](#finalizationregistryprototypeunregister)
+    * [`FinalizationRegistry.prototype.cleanupSome`](#finalizationregistryprototypecleanupsome)
+
+## FinalizationRegistry Overview
+
+A `FinalizationRegistry` instance (a "registry") lets you get *cleanup callbacks* after objects registered with the registry are reclaimed. You create the registry passing in the callback:
+
+```js
+const registry = new FInalizationRegistry(heldValue => {
```
Comment (littledan) [inline reference.md]: Fix casing: FinalizationRegistry

File: reference.md [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+* If multiple WeakRefs have the same target, they're consistent with one another. The result of calling `deref` on one of them will match the result of calling `deref` on another of them (in the same job), you won't get the target object from one of them but `undefined` from another.
+* If the target of a WeakRef is also in a [`FinalizationRegistry`](#finalizationregistry), the WeakRef's target is cleared *before* any cleanup callback associated with the registry is called.
+* You cannot change the target of a WeakRef, it will always only ever be the original target object or `undefined` when that target has been reclaimed.
+* A WeakRef might never return `undefined` from `deref`, even if nothing strongly holds the target, because the garbage collector may never decide to reclaim the object.
+
+# `FinalizationRegistry`
+
+A finalization registry provides a way to request that a *cleanup callback* get called at some point after an object registered with the registry has been *reclaimed* (garbage collected). (Cleanup callbacks are sometimes called *finalizers*.)
+
+**NOTE:** The form of this API is under review, see [issue #187](https://github.com/tc39/proposal-weakrefs/pull/187).
```
Comment (littledan) [inline reference.md]: Probably say here, explicitly, "This document is written in terms of that PR landing."

File: reference.md [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+A conforming JavaScript implementation, even one that does garbage collection, is not required to call cleanup callbacks. When and whether it does so is entirely down to the implementation of the JavaScript engine. When a registered object is reclaimed, the cleanup callbacks associated with any registries it's registered with may be called some time later, or not at all.
+
+It's likely that major implementations will call cleanup callbacks at some point during execution, but those calls may be substantially after the related object was reclaimed.
+
+There are also situations where even implementations that normally call cleanup callbacks are unlikely to call them:
+
+* When the JavaScript realm the objects existed in is being destroyed entirely (for instance, closing a tab in a browser).
+* When the `FinalizationRegistry` instance itself is no longer reachable by JavaScript code.
+
+See [Use Cases for Cleanup Callbacks](#use-cases-for-cleanup-callbacks) and [Patterns to Avoid With Cleanup Callbacks](#patterns-to-avoid-with-cleanup-callbacks) for more on this topic.
```
Comment (littledan) [inline reference.md]: These appear to be dead links. Maybe delete this paragraph for now?

File: reference.md [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+
+Developers shouldn't rely on cleanup callbacks for essential program logic. Cleanup callbacks may be useful for reducing memory usage across the course of a program, but are unlikely to be useful otherwise.
+
+A conforming JavaScript implementation, even one that does garbage collection, is not required to call cleanup callbacks. When and whether it does so is entirely down to the implementation of the JavaScript engine. When a registered object is reclaimed, the cleanup callbacks associated with any registries it's registered with may be called some time later, or not at all.
+
+It's likely that major implementations will call cleanup callbacks at some point during execution, but those calls may be substantially after the related object was reclaimed.
+
+There are also situations where even implementations that normally call cleanup callbacks are unlikely to call them:
+
+* When the JavaScript realm the objects existed in is being destroyed entirely (for instance, closing a tab in a browser).
```
Comment (littledan) [inline reference.md]: I think destruction is more of a per-agent thing than a per-realm thing. But let's just talk about this at a higher level, like, "When the JavaScript program shuts down entirely (for instance, ..."

File: reference.md [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+* If multiple WeakRefs have the same target, they're consistent with one another. The result of calling `deref` on one of them will match the result of calling `deref` on another of them (in the same job), you won't get the target object from one of them but `undefined` from another.
+* If the target of a WeakRef is also in a [`FinalizationRegistry`](#finalizationregistry), the WeakRef's target is cleared *before* any cleanup callback associated with the registry is called.
```
Comment (littledan) [inline reference.md]: before or at the same time. Maybe it'd be more clear to put this backwards, and as part of the FinalizationGroup notes: it won't be possible, during a finalizer callback, to read a WeakRef and get the object that was registered--these Weakefs will always be undefined.

Comment (tjcrowder) [reply reference.md]: I'm happy to move it. :-) But it literally is *before*, right? In [Execution](https://tc39.es/proposal-weakrefs/#sec-weakref-execution), for a given `obj`, the agent clears any [[WeakRefTarget]]s of WeakRefs before moving on to finalization for it (Step 1.a comes before Step 1.b). I find it easier to explain and reason about this if I know that when `obj` is going away / gone, any WeakRefs for it are cleared before any cleanup callbacks for it occur. The "before" here doesn't refer to event turns or anything, just the boring old temporal meaning. :-)

Comment (tjcrowder) [reply reference.md]: (I have the other fixes ready to go -- can't believe I missed a couple of those -- just waiting to resolve this point.)

Comment (littledan) [reply reference.md]: Those steps all happen "atomically", so it's not really before.

**Comment by littledan** [comment]
Thank you for the contribution!

**Comment by orta** [comment]
Nice work - https://github.com/microsoft/TypeScript/pull/38232 turns this into a d.ts file for the upcoming TypeScript 4.0 release too 👍

**Pull request merged**